# Hyper-Parameter Tuning

This notebook contains hyperparameter tuning for the model. Learner module, FastAI, provides a convenient way to create and fine-tune convolutional neural network (CNN) models. vision.learner is a function that helps us to construct a learner object, which has the model architecture, data, training configuration, and other elements. We can specify a pre-trained model architecture and fine-tune it on the dataset. vision.learner supports a wide range of CNN architectures.

This notebook includes these following implementations:

1. Random Search optimization algorithm - Run 1
2. Random Search optimization algorithm - Run 2
3. Hyperparameter Optimization with Optuna's Successive Halving Pruner

In [1]:
import warnings
warnings.filterwarnings("ignore")

In [2]:
!pip install optuna-integration

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 93.4/93.4 kB 2.5 MB/s eta 0:00:00


In [3]:
!pip install optuna lightgbm

In [4]:
!pip install --upgrade optuna

# Random Search optimization algorithm - Run 1

The optimization algorithm used below is Optuna's default algorithm, which is a Tree-structured Parzen Estimator (TPE) algorithm.

For n_trails = 10, the accuracy score and best hyper-parameter are as follows:

| Trail No. | Best Score | Architecture | Weight Decay | Epochs | Batch Size | Drop |
|-----------|------------|--------------|--------------|--------|------------|------|
| 0         | 0.9011     | ResNet34     | 0.00024      | 8      | 64         | 0.4  |
| 1         | 0.9413     | ResNet18     | 0.0090       | 15     | 64         | 0.2  |
| 2         | 0.9343     | ResNet18     | 0.0065       | 5      | 32         | 0.4  |
| 3         | 0.9080     | ResNet34     | 0.00062      | 5      | 32         | 0.4  |
| 4         | 0.9019     | ResNet34     | 0.00092      | 6      | 64         | 0.2  |
| 5         | 0.9527     | ResNet50     | 0.00005      | 7      | 64         | 0.4  |
| 6         | 0.9220     | ResNet34     | 0.00895      | 15     | 64         | 0.2  |
| 7         | 0.9404     | ResNet18     | 0.0035       | 11     | 64         | 0.4  |
| 8         | 0.9212     | ResNet18     | 0.0002       | 5      | 64         | 0.4  |
| 9         | 0.9203     | ResNet34     | 0.0007       | 13     | 32         | 0.4  |


**The best aacuracy score is 0.9527 with these hyperparameters:**
- Architecture: ResNet 50
- Weight Decay: 5.527e-5
- Epochs: 7
- Batch Size: 64
- Drop: 0.4

In [5]:
import optuna
from fastai.vision.all import *
from sklearn.model_selection import cross_val_score

# objective function for hyperparameter optimization
def objective(trial):
    path = Path('/kaggle/input/brain-tumor-mri-classification-dataset/Brain_Tumor_MRI_Image_Dataset/Training')
    
    # hyperparameters to optimize
    arch = trial.suggest_categorical('arch', ['resnet18', 'resnet34', 'resnet50'])
    wd = trial.suggest_loguniform('wd', 1e-6, 1e-2)
    epochs = trial.suggest_int('epochs', 5, 15)
    bs = trial.suggest_categorical('bs', [32, 64])
    drop = trial.suggest_categorical('drop', [0.2, 0.4])
    
    # DataBlock to prepare the data
    dblock = DataBlock(blocks=(ImageBlock, CategoryBlock),
                       get_items=get_image_files,
                       get_y=parent_label,
                       splitter=RandomSplitter(valid_pct=0.2, seed=42))
    
    dls = dblock.dataloaders(path, bs=bs) # Create DataLoaders
    
    # learner with specified architecture, metrics, weight decay, and data
    learn = vision_learner(dls, arch, metrics=accuracy, wd=wd)
    
    # Fine-tune the model
    learn.fine_tune(epochs, base_lr=0.001, cbs=[MixedPrecision()])
    
    return float(learn.validate()[1]) # validation accuracy of the trained model

# Optuna study object for hyperparameter optimization
study = optuna.create_study(direction="maximize")

# Optimize the objective function by running multiple trials
study.optimize(objective, n_trials=10)

trial = study.best_trial # best trial from the study

print("Accuracy: {}".format(trial.value))
print("Best hyperparameters: {}".format(trial.params))

[I 2024-05-24 21:50:09,794] A new study created in memory with name: no-name-10ac8b0a-751b-4f7e-b013-29bab511d22f


model.safetensors:   0%|          | 0.00/87.3M [00:00<?, ?B/s]

[I 2024-05-24 21:52:42,748] Trial 0 finished with value: 0.9010508060455322 and parameters: {'arch': 'resnet34', 'wd': 0.00023340994424944704, 'epochs': 8, 'bs': 64, 'drop': 0.4}. Best is trial 0 with value: 0.9010508060455322.


model.safetensors:   0%|          | 0.00/46.8M [00:00<?, ?B/s]

[I 2024-05-24 21:55:42,252] Trial 1 finished with value: 0.9413309693336487 and parameters: {'arch': 'resnet18', 'wd': 0.009064248812822551, 'epochs': 15, 'bs': 64, 'drop': 0.2}. Best is trial 1 with value: 0.9413309693336487.


[I 2024-05-24 21:56:56,249] Trial 2 finished with value: 0.9343257546424866 and parameters: {'arch': 'resnet18', 'wd': 0.006544389818968023, 'epochs': 5, 'bs': 32, 'drop': 0.4}. Best is trial 1 with value: 0.9413309693336487.


[I 2024-05-24 21:58:42,859] Trial 3 finished with value: 0.9080560207366943 and parameters: {'arch': 'resnet34', 'wd': 0.0006280450180845069, 'epochs': 5, 'bs': 32, 'drop': 0.4}. Best is trial 1 with value: 0.9413309693336487.


[I 2024-05-24 22:00:36,999] Trial 4 finished with value: 0.9019264578819275 and parameters: {'arch': 'resnet34', 'wd': 0.0009150842875866279, 'epochs': 6, 'bs': 64, 'drop': 0.2}. Best is trial 1 with value: 0.9413309693336487.


model.safetensors:   0%|          | 0.00/102M [00:00<?, ?B/s]

[I 2024-05-24 22:04:07,216] Trial 5 finished with value: 0.9527145624160767 and parameters: {'arch': 'resnet50', 'wd': 5.527401639726816e-05, 'epochs': 7, 'bs': 64, 'drop': 0.4}. Best is trial 5 with value: 0.9527145624160767.


[I 2024-05-24 22:08:28,633] Trial 6 finished with value: 0.9220665693283081 and parameters: {'arch': 'resnet34', 'wd': 0.008951501361629699, 'epochs': 15, 'bs': 64, 'drop': 0.2}. Best is trial 5 with value: 0.9527145624160767.


[I 2024-05-24 22:10:36,906] Trial 7 finished with value: 0.9404553174972534 and parameters: {'arch': 'resnet18', 'wd': 0.003537091430333875, 'epochs': 11, 'bs': 64, 'drop': 0.4}. Best is trial 5 with value: 0.9527145624160767.


[I 2024-05-24 22:11:41,417] Trial 8 finished with value: 0.9211909174919128 and parameters: {'arch': 'resnet18', 'wd': 0.00020852879448736512, 'epochs': 5, 'bs': 64, 'drop': 0.4}. Best is trial 5 with value: 0.9527145624160767.


[I 2024-05-24 22:15:43,586] Trial 9 finished with value: 0.9203152656555176 and parameters: {'arch': 'resnet34', 'wd': 0.0007385132863360378, 'epochs': 13, 'bs': 32, 'drop': 0.4}. Best is trial 5 with value: 0.9527145624160767.


Accuracy: 0.9527145624160767
Best hyperparameters: {'arch': 'resnet50', 'wd': 5.527401639726816e-05, 'epochs': 7, 'bs': 64, 'drop': 0.4}


# Randomized search for hyperparameter optimization - Run 2


For n_trails = 10, the accuracy score and best hyper-parameter are as follows:

| Trial | Best Score | Architecture | Weight Decay | Epochs | Batch Size | Drop |
|-------|------------|--------------|--------------|--------|------------|------|
| 0     | 0.9177     | ResNet34     | 0.000248     | 6      | 32         | 0.2  |
| 1     | 0.9492     | ResNet50     | 0.002137     | 7      | 64         | 0.4  |
| 2     | 0.9518     | ResNet50     | 0.000004     | 8      | 64         | 0.2  |
| 3     | 0.9046     | ResNet34     | 0.000269     | 6      | 64         | 0.2  |
| 4     | 0.9378     | ResNet50     | 0.000058     | 6      | 32         | 0.2  |
| 5     | 0.9352     | ResNet18     | 0.000154     | 7      | 32         | 0.2  |
| 6     | 0.9063     | ResNet34     | 0.000006     | 5      | 64         | 0.4  |
| 7     | 0.9553     | ResNet50     | 0.000004     | 13     | 64         | 0.2  |
| 8     | 0.9238     | ResNet34     | 0.000824     | 13     | 64         | 0.4  |
| 9     | 0.9361     | ResNet18     | 0.000018     | 8      | 32         | 0.2  |


**The best aacuracy score is 0.9553 with these hyperparameters:**
- Architecture: ResNet 50
- Weight Decay: 4e-6
- Epochs: 5
- Batch Size: 64
- Drop: 0.2

In [6]:
import optuna
from fastai.vision.all import *
from sklearn.model_selection import cross_val_score
from optuna.samplers import TPESampler

def objective(trial):
    path = Path('/kaggle/input/brain-tumor-mri-classification-dataset/Brain_Tumor_MRI_Image_Dataset/Training')
    
    arch = trial.suggest_categorical('arch', ['resnet18', 'resnet34', 'resnet50'])
    wd = trial.suggest_loguniform('wd', 1e-6, 1e-2)
    epochs = trial.suggest_int('epochs', 5, 15)
    bs = trial.suggest_categorical('bs', [32, 64])
    drop = trial.suggest_categorical('drop', [0.2, 0.4])

    dblock = DataBlock(blocks=(ImageBlock, CategoryBlock),
                       get_items=get_image_files,
                       get_y=parent_label,
                       splitter=RandomSplitter(valid_pct=0.2, seed=42))
    
    dls = dblock.dataloaders(path, bs=bs)

    learn = vision_learner(dls, arch, metrics=accuracy, wd=wd)
    learn.fine_tune(epochs, base_lr=0.001, cbs=[MixedPrecision()])

    return float(learn.validate()[1])

sampler = TPESampler(seed=42)  # Initialize TPESampler for random search
study = optuna.create_study(direction="maximize", sampler=sampler)  # Use TPESampler for random search
study.optimize(objective, n_trials=10)

trial = study.best_trial

print("Accuracy: {}".format(trial.value))
print("Best hyperparameters: {}".format(trial.params))

[I 2024-05-24 22:15:43,603] A new study created in memory with name: no-name-b9e79e9e-2be7-4f61-82ff-92e3b89629f2


[I 2024-05-24 22:17:45,930] Trial 0 finished with value: 0.917688250541687 and parameters: {'arch': 'resnet34', 'wd': 0.0002481040974867811, 'epochs': 6, 'bs': 32, 'drop': 0.2}. Best is trial 0 with value: 0.917688250541687.


[I 2024-05-24 22:21:15,615] Trial 1 finished with value: 0.9492118954658508 and parameters: {'arch': 'resnet50', 'wd': 0.002136832907235877, 'epochs': 7, 'bs': 64, 'drop': 0.4}. Best is trial 1 with value: 0.9492118954658508.


[I 2024-05-24 22:25:07,888] Trial 2 finished with value: 0.9518388509750366 and parameters: {'arch': 'resnet50', 'wd': 3.6138942712165278e-06, 'epochs': 8, 'bs': 64, 'drop': 0.2}. Best is trial 2 with value: 0.9518388509750366.


[I 2024-05-24 22:26:59,027] Trial 3 finished with value: 0.9045534133911133 and parameters: {'arch': 'resnet34', 'wd': 0.000269264691008618, 'epochs': 6, 'bs': 64, 'drop': 0.2}. Best is trial 2 with value: 0.9518388509750366.


[I 2024-05-24 22:30:13,466] Trial 4 finished with value: 0.9378283619880676 and parameters: {'arch': 'resnet50', 'wd': 5.762487216478604e-05, 'epochs': 6, 'bs': 32, 'drop': 0.2}. Best is trial 2 with value: 0.9518388509750366.


[I 2024-05-24 22:31:44,996] Trial 5 finished with value: 0.9352014064788818 and parameters: {'arch': 'resnet18', 'wd': 0.00015375920235481777, 'epochs': 7, 'bs': 32, 'drop': 0.2}. Best is trial 2 with value: 0.9518388509750366.


[I 2024-05-24 22:33:18,873] Trial 6 finished with value: 0.9063047170639038 and parameters: {'arch': 'resnet34', 'wd': 6.0803901902966035e-06, 'epochs': 5, 'bs': 64, 'drop': 0.4}. Best is trial 2 with value: 0.9518388509750366.


[I 2024-05-24 22:39:23,135] Trial 7 finished with value: 0.9553415179252625 and parameters: {'arch': 'resnet50', 'wd': 3.6618192203924288e-06, 'epochs': 13, 'bs': 64, 'drop': 0.2}. Best is trial 7 with value: 0.9553415179252625.


[I 2024-05-24 22:43:13,933] Trial 8 finished with value: 0.9238178730010986 and parameters: {'arch': 'resnet34', 'wd': 0.0008241925264876454, 'epochs': 13, 'bs': 64, 'drop': 0.4}. Best is trial 7 with value: 0.9553415179252625.


[I 2024-05-24 22:45:05,289] Trial 9 finished with value: 0.9360770583152771 and parameters: {'arch': 'resnet18', 'wd': 1.753594952976443e-05, 'epochs': 8, 'bs': 32, 'drop': 0.2}. Best is trial 7 with value: 0.9553415179252625.


Accuracy: 0.9553415179252625
Best hyperparameters: {'arch': 'resnet50', 'wd': 3.6618192203924288e-06, 'epochs': 13, 'bs': 64, 'drop': 0.2}


# Hyperparameter Optimization with Optuna's Successive Halving Pruner

The addition of SuccessiveHalvingPruner helps to improve efficiency during optimization. By early stopping trials with low validation accuracy (potentially bad performers), SHP focuses resources on trials that are more likely to be good. This can significantly reduce the total training time, especially for computationally expensive models.

For n_trails = 10, accuracy scores and hyper-paramters are:

| Trial | Best Score | Architecture | Weight Decay | Epochs | Batch Size | Drop               |
|-------|------------|--------------|--------------|--------|------------|--------------------|
| 0     | 0.9623     | resnet50     | 2.057e-06    | 8      | 32         | 0.2888             |
| 1     | 0.9518     | resnet50     | 0.001421     | 7      | 32         | 0.2707             |
| 2     | 0.9807     | resnet34     | 3.744e-06    | 9      | 32         | 0.3918             |
| 3     | 0.9641     | resnet18     | 2.667e-05    | 7      | 64         | 0.3196             |
| 4     | 0.9711     | resnet34     | 0.005883     | 8      | 64         | 0.2000             |
| 5     | 0.9650     | resnet18     | 5.694e-06    | 6      | 64         | 0.2266             |
| 6     | 0.9737     | resnet18     | 1.813e-05    | 6      | 64         | 0.3732             |
| 7     | 0.9851     | resnet34     | 0.004016     | 13     | 32         | 0.2680             |
| 8     | 0.9667     | resnet50     | 0.008095     | 15     | 32         | 0.3013             |
| 9     | 0.9632     | resnet50     | 0.0006995    | 6      | 32         | 0.3595             |


**The best aacuracy score is 0.9851 with these hyperparameters:**

- Architecture: ResNet 34
- Weight Decay: 0.004016
- Epochs: 13
- Batch Size: 32
- Drop: 0.268

SHA identifies and eliminates underperforming configurations early on. This prevents further resource expenditure on trials unlikely to yield good results. Optuna's Successive Halving (SHA) pruner can be advantageous over a simple random search for hyperparameter optimization.

In [8]:
from fastai.vision.all import *
import optuna
import numpy as np
from optuna.pruners import SuccessiveHalvingPruner

# objective function for Optuna
def objective(trial):
    path = Path('/kaggle/input/brain-tumor-mri-classification-dataset/Brain_Tumor_MRI_Image_Dataset/Training')

    # hyperparameters
    arch = trial.suggest_categorical('arch', [resnet18, resnet34, resnet50])
    wd = trial.suggest_loguniform('wd', 1e-6, 1e-2)
    epochs = trial.suggest_int('epochs', 5, 15)
    bs = trial.suggest_categorical('bs', [32, 64])
    drop = trial.suggest_float('drop', 0.2, 0.4)

    #DataBlock
    dblock = DataBlock(blocks=(ImageBlock, CategoryBlock),
                       get_items=get_image_files,
                       get_y=parent_label,
                       splitter=RandomSplitter(valid_pct=0.2, seed=42))
    
    dls = dblock.dataloaders(path, bs=bs)
    
    # learner
    learn = vision_learner(dls, arch, metrics=[accuracy], wd=wd)
    
    # Train the model
    learn.fine_tune(epochs, base_lr=0.001, cbs=[MixedPrecision()])
    
    # validation accuracy
    accuracy_metric = float(learn.validate()[1])
    
    return accuracy_metric

# Optuna study with the Halving Pruner
study = optuna.create_study(direction='maximize', pruner=SuccessiveHalvingPruner())

# Optimize the objective function
study.optimize(objective, n_trials=10)

# best trial
trial = study.best_trial

print("Best Accuracy: {}".format(trial.value))
print("Best hyperparameters: {}".format(trial.params))

[I 2024-05-24 23:18:32,835] A new study created in memory with name: no-name-3b9d03b9-fb10-4d8b-834d-1b4f896e975e


[I 2024-05-24 23:22:49,077] Trial 0 finished with value: 0.9623467326164246 and parameters: {'arch': <function resnet50 at 0x7a55d092f880>, 'wd': 2.056715875355261e-06, 'epochs': 8, 'bs': 32, 'drop': 0.2887557735470722}. Best is trial 0 with value: 0.9623467326164246.


[I 2024-05-24 23:26:35,485] Trial 1 finished with value: 0.9518388509750366 and parameters: {'arch': <function resnet50 at 0x7a55d092f880>, 'wd': 0.001421022419179072, 'epochs': 7, 'bs': 32, 'drop': 0.27067631749319276}. Best is trial 0 with value: 0.9623467326164246.
Downloading: "https://download.pytorch.org/models/resnet34-b627a593.pth" to /root/.cache/torch/hub/checkpoints/resnet34-b627a593.pth
100%|██████████| 83.3M/83.3M [00:01<00:00, 82.3MB/s]


[I 2024-05-24 23:29:36,291] Trial 2 finished with value: 0.9807355403900146 and parameters: {'arch': <function resnet34 at 0x7a55d092f6d0>, 'wd': 3.743608702148963e-06, 'epochs': 9, 'bs': 32, 'drop': 0.391776451145661}. Best is trial 2 with value: 0.9807355403900146.


[I 2024-05-24 23:31:07,263] Trial 3 finished with value: 0.9640980958938599 and parameters: {'arch': <function resnet18 at 0x7a55d092f520>, 'wd': 2.666590746016712e-05, 'epochs': 7, 'bs': 64, 'drop': 0.3196090658228747}. Best is trial 2 with value: 0.9807355403900146.


[I 2024-05-24 23:33:36,021] Trial 4 finished with value: 0.971103310585022 and parameters: {'arch': <function resnet34 at 0x7a55d092f6d0>, 'wd': 0.005882789424928583, 'epochs': 8, 'bs': 64, 'drop': 0.20001891036926267}. Best is trial 2 with value: 0.9807355403900146.


[I 2024-05-24 23:34:56,095] Trial 5 finished with value: 0.9649737477302551 and parameters: {'arch': <function resnet18 at 0x7a55d092f520>, 'wd': 5.693699927575653e-06, 'epochs': 6, 'bs': 64, 'drop': 0.2265821977897189}. Best is trial 2 with value: 0.9807355403900146.


[I 2024-05-24 23:36:13,018] Trial 6 finished with value: 0.9737303256988525 and parameters: {'arch': <function resnet18 at 0x7a55d092f520>, 'wd': 1.812866118696492e-05, 'epochs': 6, 'bs': 64, 'drop': 0.37324777487924343}. Best is trial 2 with value: 0.9807355403900146.


[I 2024-05-24 23:40:21,627] Trial 7 finished with value: 0.9851138591766357 and parameters: {'arch': <function resnet34 at 0x7a55d092f6d0>, 'wd': 0.004016134981130656, 'epochs': 13, 'bs': 32, 'drop': 0.2679709587877348}. Best is trial 7 with value: 0.9851138591766357.


[I 2024-05-24 23:47:49,546] Trial 8 finished with value: 0.9667250514030457 and parameters: {'arch': <function resnet50 at 0x7a55d092f880>, 'wd': 0.008095083336439402, 'epochs': 15, 'bs': 32, 'drop': 0.30129588154296516}. Best is trial 7 with value: 0.9851138591766357.


[I 2024-05-24 23:51:04,891] Trial 9 finished with value: 0.9632224440574646 and parameters: {'arch': <function resnet50 at 0x7a55d092f880>, 'wd': 0.0006994648882504557, 'epochs': 6, 'bs': 32, 'drop': 0.3595360164107316}. Best is trial 7 with value: 0.9851138591766357.


Best Accuracy: 0.9851138591766357
Best hyperparameters: {'arch': <function resnet34 at 0x7a55d092f6d0>, 'wd': 0.004016134981130656, 'epochs': 13, 'bs': 32, 'drop': 0.2679709587877348}
